# eBay Autos Analysis

## Import Libraries

In [1]:
import numpy as np
import pandas as pd

## Import eBay Autos Dataset

Below, we import the Autos dataset containing 50,000 listings from eBay Kleinanzeigen, the German eBay site. This dataset is hosted on Kaggle (https://www.kaggle.com/orgesleka/used-cars-database/data), but it has been dirtied by Dataquest.

The data dictionary is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.


After the import, we review high-level information about the dataset. Right away, we can identify items requiring clean-up:

1. Format of column titles must be changed from snakecase to camelcase
2. Some columns appear to contain numeric data (like 'price'), but they're currently stored as pandas object data type

In [2]:
# Windows
autos = pd.read_csv(r'C:\Users\Andrew\Google Drive\Colab_Notebooks\data-science-portfolio-datasets\ebay\autos.csv', encoding='Windows-1252')

In [3]:
# MacOS
#autos = pd.read_csv('~/Google_Drive/Colab_Notebooks/data-science-portfolio-datasets/ebay/autos.csv', encoding='Latin-1')

In [4]:
# Google Colab
#from google.colab import drive
#drive.mount('/content/drive')
#autos = pd.read_csv('/content/drive/My Drive/Colab_Notebooks/data-science-portfolio-datasets/ebay/autos.csv', encoding='Windows-1252')

In [5]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

## Convert Columns to SnakeCase

The original dataset includes column titles using snakecase. However, per analyst preference, we'll modify them to camelcase instead.

In [7]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [8]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [9]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

## Data Cleaning

The below section includes several data cleaning tasks, including the following:

1. Drop unnecessary columns that contain (almost the) same value for every record.
2. Covert numeric data to a true numeric data type. This requires the removal of text-based characters.
3. Identify and remove outliers

In [10]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### Drop Unnecessary Columns

The above results indicate a few columns that appear to show the same value for (almost) every row:
* seller
* offer_type
* abtest
* gearbox
* unrepaired_damage

Below, we'll evaluate each of these five columns to determine the frequency for each unique value. If almost every record is categorized as a single value, we'll drop the respective column from the dataframe.

#### Seller

In [11]:
autos.seller.unique()

array(['privat', 'gewerblich'], dtype=object)

In [12]:
autos.groupby('seller').count()

date_crawled   name  offer_type  price  abtest  vehicle_type  \
seller                                                                     
gewerblich             1      1           1      1       1             1   
privat             49999  49999       49999  49999   49999         44904   

            registration_year  gearbox  power_ps  model  odometer  \
seller                                                              
gewerblich                  1        1         1      1         1   
privat                  49999    47319     49999  47241     49999   

            registration_month  fuel_type  brand  unrepaired_damage  \
seller                                                                
gewerblich                   1          1      1                  0   
privat                   49999      45517  49999              40171   

            ad_created  nr_of_pictures  postal_code  last_seen  
seller                                                          
gewerblich           1               1            1          1  
privat           49999           49999        49999      49999

In [13]:
autos.drop('seller', axis=1, inplace=True)

#### Offer Type

In [14]:
autos.offer_type.unique()

array(['Angebot', 'Gesuch'], dtype=object)

In [15]:
autos.groupby('offer_type').count()

date_crawled   name  price  abtest  vehicle_type  \
offer_type                                                     
Angebot            49999  49999  49999   49999         44904   
Gesuch                 1      1      1       1             1   

            registration_year  gearbox  power_ps  model  odometer  \
offer_type                                                          
Angebot                 49999    47320     49999  47241     49999   
Gesuch                      1        0         1      1         1   

            registration_month  fuel_type  brand  unrepaired_damage  \
offer_type                                                            
Angebot                  49999      45518  49999              40171   
Gesuch                       1          0      1                  0   

            ad_created  nr_of_pictures  postal_code  last_seen  
offer_type                                                      
Angebot          49999           49999        49999      49999  
Gesuch               1               1            1          1

In [16]:
autos.drop('offer_type', axis=1, inplace=True)

#### AB Test

In [17]:
autos.abtest.unique()

array(['control', 'test'], dtype=object)

In [18]:
autos.groupby('abtest').count()

date_crawled   name  price  vehicle_type  registration_year  gearbox  \
abtest                                                                          
control         24244  24244  24244         21752              24244    22963   
test            25756  25756  25756         23153              25756    24357   

         power_ps  model  odometer  registration_month  fuel_type  brand  \
abtest                                                                     
control     24244  22879     24244               24244      22104  24244   
test        25756  24363     25756               25756      23414  25756   

         unrepaired_damage  ad_created  nr_of_pictures  postal_code  last_seen  
abtest                                                                          
control              19461       24244           24244        24244      24244  
test                 20710       25756           25756        25756      25756

#### Gearbox

In [19]:
autos.gearbox.unique()

array(['manuell', 'automatik', nan], dtype=object)

In [20]:
autos.groupby('gearbox').count()

date_crawled   name  price  abtest  vehicle_type  \
gearbox                                                       
automatik         10327  10327  10327   10327          9752   
manuell           36993  36993  36993   36993         33816   

           registration_year  power_ps  model  odometer  registration_month  \
gearbox                                                                       
automatik              10327     10327   9805     10327               10327   
manuell                36993     36993  35303     36993               36993   

           fuel_type  brand  unrepaired_damage  ad_created  nr_of_pictures  \
gearbox                                                                      
automatik       9796  10327               8861       10327           10327   
manuell        34247  36993              30373       36993           36993   

           postal_code  last_seen  
gearbox                            
automatik        10327      10327  
manuell          36993      36993

#### Unrepaired Damage

In [21]:
autos.unrepaired_damage.unique()

array(['nein', nan, 'ja'], dtype=object)

In [22]:
autos.groupby('unrepaired_damage').count()

date_crawled   name  price  abtest  vehicle_type  \
unrepaired_damage                                                     
ja                         4939   4939   4939    4939          4439   
nein                      35232  35232  35232   35232         33449   

                   registration_year  gearbox  power_ps  model  odometer  \
unrepaired_damage                                                          
ja                              4939     4737      4939   4617      4939   
nein                           35232    34497     35232  34038     35232   

                   registration_month  fuel_type  brand  ad_created  \
unrepaired_damage                                                     
ja                               4939       4429   4939        4939   
nein                            35232      33551  35232       35232   

                   nr_of_pictures  postal_code  last_seen  
unrepaired_damage                                          
ja                           4939         4939       4939  
nein                        35232        35232      35232

#### Explanation

Both the 'Seller' and 'Offer Type' columns contained the same record for 49,999 rows with a single outlier. Because of this, both columns were dropped from the dataframe. Alternatively, the 'Gearbox', 'AB Test', and 'Unrepaired Damage' columns contained only a couple unique values, but they were spread across tens of thousands of rows. This indicates that the columns provide meaningful value, so they were kept.

### Convert Columns to Numeric

The 'Price' and 'Odometer' columns represent numeric data, but they are stored as text. Below, we remove any non-numeric values from these columns and convert the data type.

In [23]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   price               50000 non-null  object
 3   abtest              50000 non-null  object
 4   vehicle_type        44905 non-null  object
 5   registration_year   50000 non-null  int64 
 6   gearbox             47320 non-null  object
 7   power_ps            50000 non-null  int64 
 8   model               47242 non-null  object
 9   odometer            50000 non-null  object
 10  registration_month  50000 non-null  int64 
 11  fuel_type           45518 non-null  object
 12  brand               50000 non-null  object
 13  unrepaired_damage   40171 non-null  object
 14  ad_created          50000 non-null  object
 15  nr_of_pictures      50000 non-null  int64 
 16  postal_code         50

#### Price

In [24]:
autos.price.unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [25]:
autos[autos.columns[2]] = autos[autos.columns[2]].replace('[\$,]', '', regex=True).astype(float)

#### Odometer

In [26]:
autos.odometer.unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [27]:
autos[autos.columns[9]] = autos[autos.columns[9]].replace('[\,km]', '', regex=True).astype(float)

In [28]:
autos.rename(columns = {'odometer':'odometer_km'}, inplace=True)

In [29]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        50000 non-null  object 
 1   name                50000 non-null  object 
 2   price               50000 non-null  float64
 3   abtest              50000 non-null  object 
 4   vehicle_type        44905 non-null  object 
 5   registration_year   50000 non-null  int64  
 6   gearbox             47320 non-null  object 
 7   power_ps            50000 non-null  int64  
 8   model               47242 non-null  object 
 9   odometer_km         50000 non-null  float64
 10  registration_month  50000 non-null  int64  
 11  fuel_type           45518 non-null  object 
 12  brand               50000 non-null  object 
 13  unrepaired_damage   40171 non-null  object 
 14  ad_created          50000 non-null  object 
 15  nr_of_pictures      50000 non-null  int64  
 16  post

In [30]:
autos.describe(include='all')

date_crawled         name         price abtest vehicle_type  \
count                 50000        50000  5.000000e+04  50000        44905   
unique                48213        38754           NaN      2            8   
top     2016-03-10 15:36:24  Ford_Fiesta           NaN   test    limousine   
freq                      3           78           NaN  25756        12859   
mean                    NaN          NaN  9.840044e+03    NaN          NaN   
std                     NaN          NaN  4.811044e+05    NaN          NaN   
min                     NaN          NaN  0.000000e+00    NaN          NaN   
25%                     NaN          NaN  1.100000e+03    NaN          NaN   
50%                     NaN          NaN  2.950000e+03    NaN          NaN   
75%                     NaN          NaN  7.200000e+03    NaN          NaN   
max                     NaN          NaN  1.000000e+08    NaN          NaN   

        registration_year  gearbox      power_ps  model    odometer_km  \
count        50000.000000    47320  50000.000000  47242   50000.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36993           NaN   4024            NaN   
mean          2005.073280      NaN    116.355920    NaN  125732.700000   
std            105.712813      NaN    209.216627    NaN   40042.211706   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     70.000000    NaN  125000.000000   
50%           2003.000000      NaN    105.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### Outlier Identification

Below, we evaluate two columns, Odometer and Price, to determine whether they contain outliers that need to be removed prior to analysis. Ultimately, the Price column includes a \\$99M car, as well as many \\0 cars. After evaluating these records, we drop the \\$99M car and any record with a price under $100.

#### Odometer KM

In [31]:
# Identify number of unique odometer values
autos.odometer_km.unique().shape

(13,)

In [32]:
# Describe odometer to identify glaring outliers shown in the min / max
autos.odometer_km.describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [33]:
# determine number of records for each of the 13 odometer values
autos.odometer_km.value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

#### Price

In [34]:
# Identify number of unique price values
autos.price.unique().shape

(2357,)

In [35]:
# Describe price to identify glaring outliers shown in the min / max
autos.price.describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [36]:
# determine number of records for each of the values
autos.price.value_counts().sort_index(ascending=False)

99999999.0       1
27322222.0       1
12345678.0       3
11111111.0       2
10000000.0       1
              ... 
5.0              2
3.0              1
2.0              3
1.0            156
0.0           1421
Name: price, Length: 2357, dtype: int64

In [37]:
# review high-end cars to determine whether they are outliers
autos[autos["price"] > 9999998]

date_crawled                                               name  \
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   

            price   abtest vehicle_type  registration_year    gearbox  \
2897   11111111.0     test    limousine               1973    manuell   
11137  10000000.0  control        coupe               1960    manuell   
24384  11111111.0     test          NaN               1995        NaN   
27371  12345678.0  control          NaN               2017        NaN   
39377  12345678.0  control          NaN               2018    manuell   
39705  99999999.0  control    limousine               1999  automatik   
42221  27322222.0  control    limousine               2014    manuell   
47598  12345678.0  control    limousine               2001    manuell   

       power_ps     model  odometer_km  registration_month fuel_type  \
2897         48    escort      50000.0                   3    benzin   
11137       368       NaN     100000.0                   1    benzin   
24384         0       NaN     150000.0                   0       NaN   
27371        95     punto     150000.0                   0       NaN   
39377        95       v40     150000.0                   6       NaN   
39705       224  s_klasse     150000.0                   9    benzin   
42221       163        c4      40000.0                   2    diesel   
47598       101    vectra     150000.0                   3    benzin   

                brand unrepaired_damage           ad_created  nr_of_pictures  \
2897             ford              nein  2016-03-12 00:00:00               0   
11137  sonstige_autos              nein  2016-03-29 00:00:00               0   
24384      volkswagen               NaN  2016-03-21 00:00:00               0   
27371            fiat               NaN  2016-03-09 00:00:00               0   
39377           volvo              nein  2016-03-08 00:00:00               0   
39705   mercedes_benz               NaN  2016-03-22 00:00:00               0   
42221         citroen               NaN  2016-03-08 00:00:00               0   
47598            opel              nein  2016-03-31 00:00:00               0   

       postal_code            last_seen  
2897         94469  2016-03-12 22:45:27  
11137        73033  2016-04-06 21:18:11  
24384        18519  2016-03-21 14:40:18  
27371        96110  2016-03-09 15:45:47  
39377        14542  2016-04-06 23:17:31  
39705        73525  2016-04-06 05:15:30  
42221        76532  2016-03-08 20:39:05  
47598         4356  2016-03-31 18:56:54

In [38]:
# review low-end cars to determine whether they are outliers
autos[autos["price"].between(10,100)]

date_crawled                                               name  \
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
64     2016-04-05 07:36:19                 Autotransport__Abschlepp_Schlepper   
245    2016-03-30 12:45:44                                     Renault_Twingo   
461    2016-03-17 12:54:04                            Gut_erhaltene_Alufelgen   
...                    ...                                                ...   
48662  2016-03-31 14:37:38  Golf_II_GL_2_tuerig___alltagstauglich___TÜV_ab...   
48795  2016-03-14 19:50:28         Golf_3_1_9l_Diesel_64_PS_zum_Ausschlachten   
49187  2016-03-20 13:43:20                            B_Corsa_Bastlerfahrzeug   
49843  2016-04-01 08:56:18                                    Mazda_323_F_1.4   
49874  2016-04-02 18:25:23                              Ford_KA_mit_Tuev_2017   

       price   abtest vehicle_type  registration_year  gearbox  power_ps  \
25      90.0  control        kombi               1996  manuell       116   
30      80.0     test   kleinwagen               2002  manuell        60   
64      40.0     test          NaN               2011      NaN         0   
245     50.0  control   kleinwagen               1997  manuell        60   
461    100.0     test          NaN               2008      NaN         0   
...      ...      ...          ...                ...      ...       ...   
48662  100.0  control          NaN               1990  manuell        71   
48795  100.0  control       andere               1991  manuell        64   
49187  100.0     test   kleinwagen               1996  manuell        45   
49843   60.0     test   kleinwagen               1998  manuell        73   
49874  100.0     test   kleinwagen               1999  manuell         0   

         model  odometer_km  registration_month fuel_type           brand  \
25         NaN     150000.0                   4    benzin            ford   
30     2_reihe     150000.0                   6    benzin         peugeot   
64         5er     150000.0                   5       NaN             bmw   
245     twingo      30000.0                   5    benzin         renault   
461        NaN      30000.0                   0       NaN  sonstige_autos   
...        ...          ...                 ...       ...             ...   
48662     golf     150000.0                   7    benzin      volkswagen   
48795     golf     150000.0                   7    diesel      volkswagen   
49187    corsa     150000.0                  12    benzin            opel   
49843  3_reihe     150000.0                   3    benzin           mazda   
49874       ka     150000.0                   6    benzin            ford   

      unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
25                   ja  2016-03-21 00:00:00               0        27574   
30                   ja  2016-03-14 00:00:00               0        57076   
64                  NaN  2016-04-05 00:00:00               0        40591   
245                 NaN  2016-03-30 00:00:00               0        31249   
461                 NaN  2016-03-17 00:00:00               0        66440   
...                 ...                  ...             ...          ...   
48662                ja  2016-03-31 00:00:00               0        83700   
48795              nein  2016-03-14 00:00:00               0        79348   
49187                ja  2016-03-20 00:00:00               0        30880   
49843               NaN  2016-04-01 00:00:00               0        34317   
49874                ja  2016-04-02 00:00:00               0        55131   

                 last_seen  
25     2016-04-01 05:16:49  
30     2016-03-14 11:47:31  
64     2016-04-07 12:16:01  
245    2016-04-07 03:16:36  
461    2016-03-27 00:16:38  
...                    ...  
48662  2016-03-31 14:37:38  
48795  2016-03-28 06:16:02  
49187  2016-03-20

In [39]:
# remove outlier - record with $99M price
autos = autos[autos["price"].between(100,30000000)]

### Date Formatting

#### Date Crawled

The code below describes a distribution of dates on which the respective data was crawled from the source website. The results indicate that the data was crawled over about a month period from March 5, 2016 through April 7, 2016.

In [40]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025354
2016-03-06    0.014035
2016-03-07    0.036051
2016-03-08    0.033211
2016-03-09    0.033024
2016-03-10    0.032278
2016-03-11    0.032589
2016-03-12    0.036922
2016-03-13    0.015673
2016-03-14    0.036652
2016-03-15    0.034310
2016-03-16    0.029459
2016-03-17    0.031511
2016-03-18    0.012895
2016-03-19    0.034724
2016-03-20    0.037793
2016-03-21    0.037233
2016-03-22    0.032900
2016-03-23    0.032278
2016-03-24    0.029438
2016-03-25    0.031490
2016-03-26    0.032299
2016-03-27    0.031117
2016-03-28    0.034952
2016-03-29    0.034144
2016-03-30    0.033729
2016-03-31    0.031864
2016-04-01    0.033688
2016-04-02    0.035595
2016-04-03    0.038601
2016-04-04    0.036569
2016-04-05    0.013061
2016-04-06    0.003172
2016-04-07    0.001389
Name: date_crawled, dtype: float64

#### Ad Created

The code below describes a distribution of dates on which the ad was created by the end user. Because they spanned over abbout a year, the corresponding daily result-set is difficult to analyze. The second calculation groups the distribution based on month. We can see that the overwhelming majority of ads were created in March 2016.

In [41]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038850
2016-04-04    0.036922
2016-04-05    0.011796
2016-04-06    0.003255
2016-04-07    0.001244
Name: ad_created, Length: 76, dtype: float64

In [42]:
autos['ad_created'].str[:7].value_counts(normalize=True, dropna=False).sort_index()

2015-06    0.000021
2015-08    0.000021
2015-09    0.000021
2015-11    0.000021
2015-12    0.000041
2016-01    0.000249
2016-02    0.001265
2016-03    0.837345
2016-04    0.161017
Name: ad_created, dtype: float64

#### Last Seen

The code below describes the last time at which the web crawler last saw the corresponding ad on the website. The distribution skews toward the last few days, indicating that ads stayed on the website through the end of the crawling period.

In [43]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001078
2016-03-06    0.004312
2016-03-07    0.005432
2016-03-08    0.007339
2016-03-09    0.009598
2016-03-10    0.010635
2016-03-11    0.012397
2016-03-12    0.023799
2016-03-13    0.008873
2016-03-14    0.012625
2016-03-15    0.015859
2016-03-16    0.016440
2016-03-17    0.028090
2016-03-18    0.007318
2016-03-19    0.015756
2016-03-20    0.020648
2016-03-21    0.020565
2016-03-22    0.021353
2016-03-23    0.018575
2016-03-24    0.019757
2016-03-25    0.019093
2016-03-26    0.016668
2016-03-27    0.015548
2016-03-28    0.020855
2016-03-29    0.022307
2016-03-30    0.024691
2016-03-31    0.023841
2016-04-01    0.022846
2016-04-02    0.024898
2016-04-03    0.025126
2016-04-04    0.024525
2016-04-05    0.125070
2016-04-06    0.221967
2016-04-07    0.132118
Name: last_seen, dtype: float64

#### Registration Year

The code below describes the Registration Year data. The results indicate a minimum year of 1000 and maximum year of 9999, indicating that incorrect data exists (as cars were not invented in 1000 AD and we have not yet entered year 9999).

In [44]:
autos['registration_year'].describe()

count    48237.000000
mean      2004.728652
std         87.886096
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [45]:
# determine outliers - cars with registration year before 1900
autos[autos['registration_year'] < 1900]

date_crawled                                   name   price  \
10556  2016-04-01 06:02:10                             UNFAL_Auto   450.0   
22316  2016-03-29 16:56:41  VW_Kaefer.__Zwei_zum_Preis_von_einem.  1500.0   
24511  2016-03-17 19:45:11            Trabant__wartburg__Ostalgie   490.0   
32585  2016-04-02 16:56:39                             UNFAL_Auto   450.0   
49283  2016-03-15 18:38:53                             Citroen_HY  7750.0   

        abtest vehicle_type  registration_year  gearbox  power_ps   model  \
10556  control          NaN               1800      NaN      1800     NaN   
22316  control          NaN               1000  manuell         0  kaefer   
24511  control          NaN               1111      NaN         0     NaN   
32585  control          NaN               1800      NaN      1800     NaN   
49283  control          NaN               1001      NaN         0  andere   

       odometer_km  registration_month fuel_type       brand  \
10556       5000.0                   2       NaN  mitsubishi   
22316       5000.0                   0    benzin  volkswagen   
24511       5000.0                   0       NaN     trabant   
32585       5000.0                   2       NaN  mitsubishi   
49283       5000.0                   0       NaN     citroen   

      unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
10556              nein  2016-04-01 00:00:00               0        63322   
22316               NaN  2016-03-29 00:00:00               0        48324   
24511               NaN  2016-03-17 00:00:00               0        16818   
32585              nein  2016-04-02 00:00:00               0        63322   
49283               NaN  2016-03-15 00:00:00               0        66706   

                 last_seen  
10556  2016-04-01 09:42:30  
22316  2016-03-31 10:15:28  
24511  2016-04-07 07:17:29  
32585  2016-04-04 14:46:21  
49283  2016-04-06 18:47:20

In [46]:
# determine outliers - cars with registration year after 2017 (the year after the data was obtained)
autos[autos['registration_year'] > 2017]

date_crawled                                               name  \
84     2016-03-27 19:52:54                                     Renault_twingo   
164    2016-03-13 20:39:16  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...   
390    2016-03-25 12:59:06   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19   
453    2016-03-28 13:51:12                                         Armee_Jeep   
802    2016-03-19 11:37:23                 Lada_mit_wenig_km_neuem_Tuev_bj_08   
...                    ...                                                ...   
49262  2016-03-26 18:31:21                                        Opel_vectra   
49354  2016-04-05 10:31:29                        Bmw_e39_523i_mit_neuem_Tuev   
49411  2016-03-28 12:58:19                            Renault_twingo_Tuev_neu   
49770  2016-03-15 12:54:26                     VW_Polo_6n_Tuev_Neu!__1.6_75PS   
49910  2016-04-03 21:39:15  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...   

         price   abtest vehicle_type  registration_year    gearbox  power_ps  \
84       900.0  control          NaN               2018        NaN        60   
164     4800.0  control          NaN               2018    manuell         0   
390     7750.0     test          NaN               2018    manuell        76   
453     9800.0     test          NaN               4500    manuell         0   
802     2100.0     test          NaN               2018    manuell         0   
...        ...      ...          ...                ...        ...       ...   
49262   2000.0     test          NaN               2018        NaN         0   
49354   2499.0  control          NaN               2018    manuell       174   
49411   1550.0     test          NaN               2018        NaN         0   
49770    999.0  control          NaN               2018    manuell        75   
49910  22200.0     test          NaN               9000  automatik       140   

        model  odometer_km  registration_month fuel_type       brand  \
84     twingo     150000.0                   0       NaN     renault   
164    meriva      80000.0                   4    benzin        opel   
390    andere     150000.0                   6    benzin        fiat   
453    andere       5000.0                   0       NaN        jeep   
802    kalina     150000.0                   0    benzin        lada   
...       ...          ...                 ...       ...         ...   
49262     NaN     150000.0                   0       NaN        opel   
49354     5er     150000.0                   8       NaN         bmw   
49411  twingo     100000.0                   0       NaN     renault   
49770    polo     150000.0                  12    benzin  volkswagen   
49910  andere      10000.0                   3    benzin        opel   

      unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
84                  NaN  2016-03-27 00:00:00               0        40589   
164                nein  2016-03-13 00:00:00               0        37627   
390                nein  2016-03-25 00:00:00               0        78239   
453                 NaN  2016-03-28 00:00:00               0         7545   
802                 NaN  2016-03-19 00:00:00               0        12621   
...                 ...                  ...             ...          ...   
49262               NaN  2016-03-26 00:00:00               0        65549   
49354              nein  2016-04-05 00:00:00               0        65207   
49411               NaN  2016-03-28 00:00:00               0        48739   
49770              nein  2016-03-15 00:00:00               0        24321   
49910               NaN  2016-04-03 00:00:00               0        56856   

                 last_seen  
84     2016-04-05 18:46:49  
164    2016-04-04 16:48:02  
390    2016-03-28 12:16:50  
453    2016-04-06 17:45:49  
802    2016-03-21 15:20:00  
...                    ...  
49262  2016-04-06 18:19:01  
49354  2016-04-05 12:27:15  
49411  2016-03-28 13:41:18  


In [47]:
# remove outliers
autos = autos[autos['registration_year'].between(1900,2017)]

In [48]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):  # more options can be specified also
    print(autos['registration_year'].value_counts(normalize=True).sort_index())

1910    0.000042
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000084
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000021
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000084
1957    0.000042
1958    0.000084
1959    0.000126
1960    0.000482
1961    0.000126
1962    0.000084
1963    0.000168
1964    0.000251
1965    0.000356
1966    0.000461
1967    0.000545
1968    0.000545
1969    0.000398
1970    0.000775
1971    0.000545
1972    0.000691
1973    0.000503
1974    0.000503
1975    0.000377
1976    0.000440
1977    0.000461
1978    0.000880
1979    0.000712
1980    0.001696
1981    0.000586
1982    0.000859
1983    0.001068
1984    0.001068
1985    0.001969
1986    0.001487
1987    0.001508
1988    0.002786
1989    0.003581
1990    0.006953
1991    0.007079
1992    0.007707
1993    0.008796
1994    0.013111
1995    0.025007
1996    0.028379
1997    0.0403

### Aggregation

In [76]:
# Determine percentage of cars in dataset per brand
brand_series = autos.brand.value_counts(normalize=True)
brand_series

volkswagen        0.212788
bmw               0.108949
opel              0.108007
mercedes_benz     0.096048
audi              0.086246
ford              0.069805
renault           0.047605
peugeot           0.029593
fiat              0.025824
seat              0.018598
skoda             0.016169
nissan            0.015373
mazda             0.015289
smart             0.014326
citroen           0.014137
toyota            0.012755
hyundai           0.009969
sonstige_autos    0.009425
volvo             0.009069
mini              0.008734
mitsubishi        0.008105
honda             0.007938
kia               0.007121
alfa_romeo        0.006660
suzuki            0.005864
porsche           0.005843
chevrolet         0.005613
chrysler          0.003519
dacia             0.002681
daihatsu          0.002492
jeep              0.002241
land_rover        0.002073
subaru            0.002073
saab              0.001634
daewoo            0.001550
jaguar            0.001487
trabant           0.001361
r

In [77]:
# Create pandas series of only the brands with greater than 10% share
brands_over_10_percent = brand_series[brand_series > 0.1]
brands_over_10_percent

volkswagen    0.212788
bmw           0.108949
opel          0.108007
Name: brand, dtype: float64

In [79]:
# Create and load a dictionary of the average price for each of the top brands
price_dict = {}

for cur_brand in brands_over_10_percent.index:
    temp_autos = autos[autos['brand'] == cur_brand]
    price_mean = temp_autos['price'].mean()
    price_dict[cur_brand] = price_mean

price_dict

{'volkswagen': 6673.283562992126,
 'bmw': 8567.470203767782,
 'opel': 5376.428931549351}

In [78]:
# Create and load a dictionary of the average registration year for each of the top brands
reg_dict = {}

for cur_brand in brands_over_10_percent.index:
    temp_autos = autos[autos['brand'] == cur_brand]
    reg_mean = temp_autos['registration_year'].mean()
    reg_dict[cur_brand] = reg_mean

reg_dict

{'volkswagen': 2002.9902559055117,
 'bmw': 2003.2954632833525,
 'opel': 2002.793872406438}

In [81]:
# Convert dictionaries to pandas series
price_series = pd.Series(price_dict)
reg_series = pd.Series(reg_dict)

In [86]:
# Create pandas dataframe combining the two series
frame = {'avg_price': price_series, 'avg_reg_year': reg_series}
aggregate_autos = pd.DataFrame(frame)
aggregate_autos

avg_price  avg_reg_year
volkswagen  6673.283563   2002.990256
bmw         8567.470204   2003.295463
opel        5376.428932   2002.793872